wire-runtime/src/main/java/com/squareup/wire/Message.java

```diff
@@ -301,4 +261,244 @@ protected IllegalStateException missingRequiredFields(Object... args) {
+      if (value == null) return;
+      if (builder.length() > 0) builder.append(", ");
+      builder.append(name).append("=").append(redacted ? "██" : value);
+    }
+
+    @Override public <T> void repeated(int tag, String name, List<T> list, TypeAdapter<T> adapter,
+        boolean redacted) {
+      value(tag, name, list, null, redacted);
+    }
+
+    @Override public <T> void packed(int tag, String name, List<T> list, TypeAdapter<T> adapter,
+        boolean redacted) {
+      value(tag, name, list, null, redacted);
+    }
+
+    @Override public void unknowns(Message<?> message) {
+    }
+
+    @Override public void extensions(ExtendableMessage<?> message) {
+      if (builder.length() > 0) builder.append(", ");
+      ExtensionMap<?> extensionMap = message.extensionMap;
+      builder.append("{extensions=")
+          .append(extensionMap == null ? "{}" : extensionMap.toString())
+          .append('}');
+    }
+
+    String finish(Message message) {
+      builder.insert(0, message.name + '{');
+      return builder.append('}').toString();
+    }
+  }
+
+  private static final class HashCodeVisitor implements Visitor {
+    int result = 0;
+
+    @Override
+    public <T> void value(int tag, String name, T value, TypeAdapter<T> adapter, boolean redacted) {
+      result = result * 37 + (value != null ? value.hashCode() : 0);
+    }
+
+    @Override public <T> void repeated(int tag, String name, List<T> list, TypeAdapter<T> adapter,
+        boolean redacted) {
+      result = result * 37 + list.hashCode();
+    }
+
+    @Override public <T> void packed(int tag, String name, List<T> list, TypeAdapter<T> adapter,
+        boolean redacted) {
+      result = result * 37 + list.hashCode();
+    }
+
+    @Override public void unknowns(Message<?> message) {
+    }
+
+    @Override public void extensions(ExtendableMessage<?> message) {
+      ExtensionMap<?> extensionMap = message.extensionMap;
+      result = result * 37 + (extensionMap != null ? extensionMap.hashCode() : 0);
+    }
+  }
+
+  private static final class EqualsVisitor implements Visitor {
+    private final List<Object> instances = new ArrayList<Object>();
+
+    @Override
+    public <T> void value(int tag, String name, T value, TypeAdapter<T> adapter, boolean redacted) {
+      instances.add(value);
+    }
+
+    @Override public <T> void repeated(int tag, String name, List<T> list, TypeAdapter<T> adapter,
+        boolean redacted) {
+      instances.add(list);
+    }
+
+    @Override public <T> void packed(int tag, String name, List<T> list, TypeAdapter<T> adapter,
+        boolean redacted) {
+      instances.add(list);
+    }
+
+    @Override public void unknowns(Message<?> message) {
+    }
+
+    @Override public void extensions(ExtendableMessage<?> message) {
+      instances.add(message.extensionMap);
+    }
+
+    boolean isEqualTo(EqualsVisitor other) {
+      return instances.equals(other.instances);
```

Comment:
Probably quite inefficient, since each nested message's `equals()` call is going to allocate two lists. Possible to implement this recursively with a single list?